### DISCUSSION TITLE
어제 회의에서 지나간 Git Hook에 대해서 개인적인 느낌을 남겨봅니다!

### DISCUSSION THREAD

**Discussion by thornscript** [GU-99/grow-up-pms#7] [개발 토론]
## 🔍 Git Hook을 통한 강제 검사
이전 회의에서 재현님을 통해 간접적으로 Git hook의 존재를 알게 되었고, 이런 훌륭한 기능을 통해 아래와 같은 부분을 강제할 수 있었으면 좋겠다고 생각했습니다.

- 커밋 전 커밋 메시지 형식 검사(pre-commit)
- 커밋 전 코드 스타일 및 포맷 검사(pre-commit, checkstyle)
- 푸시 전 모든 단위 테스트를 통과하는지 검사(pre-push)

Git hook은 개발 워크플로우에서 유용한 역할을 할 수 있을 것 같았고, 특히, 팀 내에서 코드 일관성과 품질을 유지하는 데 큰 도움이 될 수 있으리라 생각합니다. 하지만, 조사 후에는 결국 개인의 재량에 맡기는 것이 가장 좋겠다는 생각이 들었습니다. 🥹 

### 설정의 번거로움과 유지보수
Git hook 스크립트를 사용하려면 해당 스크립트가 `<project>/.git/hooks/`에 위치해야 합니다. 그러나 `.git` 디렉토리는 보통 `.gitignore`에 포함되어 변경이 추적되지 않습니다. 심볼릭 링크나 `git config core.hooksPath`를 통해 경로를 명시적으로 지정할 수 있지만, 팀원이 클론을 받은 후 수동으로 설정해야 하는 번거로움이 있을 것 같았습니다! 팀 전체에서 일관된 설정을 유지하는 데 어려움이 있을 수도 있을 것 같다는 생각이 들었어요.

### 강제성의 부족
클라이언트 사이드에서 실행되는 스크립트는 개발자에게 강제할 수 없다는 사실도 크게 다가왔습니다! 개발자가 클론 후에 별다른 설정(심볼릭 링크 생성, 훅 스크립트 경로 지정 등)을 해주지 않으면 적용이 되지 않는 부분과, 조사를 해보니 개발자의 선호도나 개발자가 Git에 익숙한지에 따라서 Git hook을 선뜻 사용하는 케이스와 그렇지 않은 케이스가 꽤 있다는 것을 알게 되었습니다 😢 

> 저는 pre-commit hook을 사용하고 있습니다. 이 기능은 매우 유용하지만, 모든 개발자가 이를 설치하는 데 동의해야 한다는 전제가 있습니다. 실제로 pre-commit hook을 설정하지 않고 프로젝트에 참여한 개발자들이 있어서, 이들이 해당 검사를 거치지 않고 커밋하는 경우가 있었습니다.
> Git 서버 수준에서 pre-commit hook을 설치하여 푸시할 때 강제하는 방법이 있는지 잘 모르겠습니다. 이와 관련된 정보를 아는 분이 있다면 피드백을 받고 싶습니다. 어쨌든, 위와 같은 상황이 발생하지 않도록 주의해야 합니다. 

>> 맞아요, 솔직히 말해서 각 프로젝트마다 '설치하지 않는 사람'이 꼭 한 명씩은 있게 마련입니다. "난 20년 동안 프로그래밍을 해왔는데, 이런 건 한 번도 필요 없었어"라고 말하는 사람이죠.
>>
>> 저는 이런 문제를 피하기 위해 서버 쪽(GitHub Actions 같은)에서 처리하는 것을 선호합니다. 이렇게 하면 아무도 이를 방해할 수 없거든요.

### 성능과 효율성 문제
이것은 자칫 저희 프로젝트와는 거리가 다소 먼 이야기일 수도 있지만, 그래도 소개해보겠습니다!

개발자는 작고 빈번하게 커밋을 올리게 되는데, pre-commit hook이 시간을 잡아먹는 경우 팀원에게 스트레스를 가중시킬 수 있다고 생각이 들었습니다. 특히, 대규모 코드 베이스에서는 린터나 스타일 검사와 같은 작업은 상당한 시간이 소요될 수 있다고 합니다. 만약, 그런게 지속된다면 개발 속도를 저하시킬 뿐만 아니라 팀원들의 작업 흐름을 방해할 수 있다는 생각이 들었습니다.

### CI 파이프라인의 장점
무료로 제공되는 CI 파이프라인을 통해 모든 필수 검사를 수행할 수 있기도 하고, 구축 비용도 낮습니다! 이 방법은 팀원이 별다른 설정을 할 필요 없이 팀 전체에 일관된 검사를 적용할 수 있어 좋다고 생각했습니다. CI 파이프라인에는 아래와 같은 이점들이 있습니다!

- **자동화된 검사**: 린터, 스타일 가이드 검사, 코드 퀄리티 검사 등을 자동으로 수행해서 일관된 코드 품질을 유지할 수 있습니다.
- **중앙 집중화된 설정**: 모든 검사를 중앙에서 설정하고 관리할 수 있어서, 각 개발자가 개별적으로 설정할 필요가 없습니다!

저는 SonarLint와 CheckStyle-IDEA를 통해 이미 실시간으로 컨벤션이나 코드 스멜에 대한 피드백을 받고 있습니다. 처음에는 Git hook을 통해 1차적으로 검사를 하는 것이 좋겠다고 생각했으나, 개발자 커뮤니티에서도 호불호가 갈리는 기능인 만큼, 플러그인을 선택할지 스크립트를 선택할지, 최후의 보루로 CI 파이프라인을 믿을지는 개인의 재량에 맡기는 것이 좋을 것 같다고 결론을 내렸습니다.

### 플러그인을 통한 즉각적인 피드백의 장점

즉각적인 피드백을 원할 때는 플러그인이 더 적합할 수 있다는 생각이 들었습니다. IDE나 텍스트 에디터에 통합되어 있는 플러그인은 코드 작성 중 실시간으로 팀원에게 피드백을 제공해서 개발자가 즉시 문제를 수정할 수 있도록 도와주는 친구들입니다. 이런 플러그인들은 팀원의 작업 흐름을 크게 방해하지도 않고, 코드 품질을 유지하는 데 큰 도움이 되는 것 같습니다. 또한, 플러그인은 설정이나 유지보수도 간편해서 각 개발자가 자신의 개발 환경에 맞게 쉽게 적용할 수도 있습니다!

## 🔚 결론
Git hook은 특정 규칙을 강제하고 코드 품질을 유지하는 데 유용하다고 생각하지만, 설정과 유지보수의 번거로움, 강제성 부족, 성능 문제 등으로 인해 팀 전체에 적용하기에는 한계가 있다는 생각이 들었습니다! 그 외의 방법으로는 CI 파이프라인을 통해 중앙에서 일관된 검사를 수행하고 좀 더 자동화된 방식으로 코드 품질을 유지하거나, 즉각적인 피드백이 필요한 경우에는 플러그인을 사용하는 것이 더 직관적이고 유용할 것 같다고 생각이 들었습니다 🐱 (어디까지나 개인적인 느낌이었습니다...!)

결국, 제가 내린 결론은 Git hook, 플러그인 중 어느 걸 택할지는 각 팀원의 재량에 맡기면 좋을 것 같다는 생각이 들었습니다.

**Comment by Seok93** [comment]
도현님이 말씀하신 대로 .git 안에 있는 파일이라 수동적으로 적용해야하고, git 파일 추적이 동작하지 않는 점이 문제여서, 예전에 했던 프로젝트에서는 pre-commit 파일을 상위로 이동시키고, 이 파일을 자동으로 구성해주는 쉘스크립트를 작성한 기억이 있네요. 

파일 수정이 있을 시 쉘스크립트를 통해 자동으로 복사하고 구성하는 방법을 사용해서 일부분은 편해졌으나, 변경사항이 있을 때 팀원들에게 알려서 다시 설정하도록 하는 등의 번거로움이 있었습니다. 

그래서 이전에 pre-commit을 설정할 때엔 코드 포맷팅, 메세지 포맷팅 등 간단한 처리만 넣고, 변경을 최소화한 방향으로 사용했어요. 코드 포맷팅이 깨진 상태로 git에 push 하면 불필요한 변경사항을 기록에 남기게 되고, 다시 수정 사항을 올리는 번거로움도 생기니 이전에 차단하는건 괜찮은 것 같아요.

결론적으로 도현님의 의견대로 테스트 검수, 빌드 검수 등은 CI 파이프로 구성하면 좋을 것 같고, 코드 포맷팅과 같은 부분은 pre-commit으로 빼면 좋지 않을까요?

**Reply by thornscript** [reply]
@Seok93 아니... 순간 재현님이 아니라 우석님이 바로 확인하실 줄은 몰랐씁니다.... ㅋㅋㅋㅋㅋㅋㅋ 백엔드 회의 도중에 pre-commit이란 얘기가 나와서 개인적으로 조사해보다가 이미 플러그인을 통해서 커밋 전에 경고를 받고 있었어서 뭔가 제 환경에서는 이중으로 검사하는 느낌이라 이 부분에 대해서는 개인의 재량에 맡기는 것도 좋겠다는 느낌을 한 번 적어봤습니다... 🥹 

Git hook 스크립트를 통해 자동화된 검사를 수행하면 사실 CI 파이프라인이 돌아가는 것도 소중한 서버 자원이고 상대적으로 빠른 피드백을 받을 수 있다는 부분이 장점인 것 같아요! Husky라는 우아한 툴도 있는 것 같지만 자바스크립트 기반 프로젝트에서는 괜찮은 선택인 것 같더라고요! 프론트 프로젝트를 구경하면서 프로젝트 루트 경로에 `.husky`를 보고 이미 적용하시고 계시는 구나라고 깨달았습니다... 👍 

<img src="https://github.com/GU-99/grow-up-pms/assets/8318212/1b4063c3-0852-4c86-834a-be528b306608" style="margin: auto; display: block;" />

사실 1차적으로 개발자 환경에서 바로 걸러낼 수 있다면 선택할 수 있는 안전 장치가 여럿 있어도 괜찮다는 생각이 들었습니다. 개발자가 빠르게 피드백을 받는 것이 무엇보다 중요하니까요! 그리고 리뷰어의 시간은 언제나 소중합니다...

**Reply by Seok93** [reply]
플로그인으로 벌써 야무지게 통제하고 계셨군요! ㅋㅋㅋㅋㅋㅋㅋ 

도현님께서 여러 방면으로 고민하시고, 개념을 확장해주셔서 백엔드쪽 이슈, 토론 등을 구경하면 재미있는 부분이 많더라고요! 가끔 놀러오겠습니다 ㅋㅋㅋㅋ

**Comment by KimJayhyun** [comment]
Git hook과 Pre commit에 대해 정리를 너무 잘해주셔서 다른 자료를 안 읽어봐도 이해가 너무 잘 됩니다!

말씀하신거와 같이 클라이언트 사이드의 pre-commit과 Git hook에는 여러 부작용이 있고 제 생각에는 아래 사항들이 아주 큰 단점인거 같네요.

### pre-commit 라이브러리의 단점
#### 강제성 부족
Git Hook을 사용하는 이유인것 같습니다. 다른 pre-commit 관련 라이브러리를 사용하기 위해선 추가적인 설치와 설정가 필요한 상황이고 항상 말을 안 듣는 사람이 있기 마련이죠...
Git Hook은 추가적인 설치나 설정이 필요없는 대신에 공유가 어렵다는 점이 아쉽네요.

#### 성능과 효율성 문제
이 문제도 심각할 것 같습니다. 개발의 효율성과 피로도를 올린다면 굳이 pre-commit 라이브러리를 도입하기 보다는 개발자 개인이 테스트를 진행하는 것이 좋아보입니다.

#### 결론적으로는 
pre-commit을 도입해야한다면 클라이언트 사이드에서의 실행은 개발자에게 자유를 주고(라이브러리를 사용하던 IDE에 플러그인을 사용하던)
CI를 활용하여 도입을 하는 것이 맞을 것 같습니다. 

어떤 방법으로 이를 도입해야 할지 더 생각해보겠습니다~

**Reply by thornscript** [reply]
@KimJayhyun 개발자 환경에서 1차적으로 거르고 CI 파이프라인을 통해서 테스트 통과 여부를 2차적으로 검증하는 게 좋아보이는 것 같은데, Git hook의 적용에 대해서는 긍정적이나 약간 개발자마다 의견 차이가 있는 것 같습니다! 프론트 쪽은 자바스크립트 기반 프로젝트인 경우 Husky처럼 사전 준비를 마쳐두면 한 번에 내려받고 설정도 간편한 것 같은 유용한 툴들이 많아서 엄청 괜찮은 선택지인 것 같더라구요... 🥹

자바 쪽에서는 그렇다 할 툴이 그렇게 보이지는 않아서 추후에 재현님이랑 로컬 환경에서 컨벤션을 강제할 수 있는 방법을 논의해보고 결정해야 겠습니다. 두 분 다 소중한 시간 내주셔서 감사합니다 😸
